README.md
# puppet-bind

[![Build Status](https://travis-ci.org/onpuppet/puppet-bind.png?branch=master)](https://travis-ci.org/onpuppet/puppet-bind)

## Overview
  
Install and enable a BIND DNS server, manage its main configuration and install
and manage its DNS zone files.

# Example bind master slave setup

```puppet
node master {
	class { 'bind': 
	          allow_notify      => [ '#{slave_ipv4}' ] 
	        }
	        
	        bind::zone { 'example.com':
	          nameservers    => ['ns1.example.com', 'ns2.example.com'],
	          allow_transfer => [ '#{slave_ipv4}' ],
	        }
	  
	        bind::zone { '12.168.192.IN-ADDR.ARPA':
	          nameservers    => ['ns1.example.com', 'ns2.example.com'],
	          allow_transfer => [ '#{slave_ipv4}' ],
	        }
	        
	        bind::record::a {
	          'gateway':
	            zone => 'example.com',
	            data => '192.168.12.1',
	            ptr  => true
	        }
	        
	        bind::record::a {
	          'mail':
	            zone => 'example.com',
	            data => '192.168.12.3',
	            ptr  => true
	        }
	  }
}

node slave {
          class { 'bind': 
            masters => { 'masterlist' => [ '#{master_ipv4}' ] }
          }
}
```


## TESTING

    # Ubuntu 14.04
    apt-get install libxslt-dev bundler -y
    bundle install
    curl -sSL https://get.docker.io/ubuntu/ | sudo sh
    sudo usermod -G docker jenkins
    bundle exec rake test
    bundle exec rake acceptance
    


CHANGELOG.md
## 2017-02-22 Release 1.0.0

* Initial public release


lib/spec/dns.rb
require 'rubygems'
require 'resolv'
require 'ipaddress'

class DNS
  attr_reader :role, :ip_address

  # ip_address = dotted notation as a string, e.g. '192.168.1.1'
  # role = 'Master' or 'Slave'
  # domain_name = the domain name as a string, e.g. 'sharknet.us'
  def initialize(ip_address, role, domain_name)
    @role = role
    @ip_address = ip_address
    @resolver = Resolv::DNS.new(
      nameserver: [ip_address],
      search: [domain_name],
      ndots: 1
    )
    @resolver.timeouts = 2 # Enable for Ruby 2.1 or newer. Fail if lookup takes longer than 2 seconds
  end

  # returns the DNS server as a string in the form '192.168.1.1 [Master]'
  def to_s
    "#{ip_address} [#{role}]"
  end

  # returns the ip address as a string or 'Hostname not found' if there was any error
  def address(hostname)
    @resolver.getaddress(hostname).to_s
  rescue
    'Hostname not found'
  end

  # returns the hostname as a string or 'IP address not found' if there was any error
  def hostname(ip_address)
    @resolver.getname(ip_address).to_s
  rescue
    'IP address not found'
  end

  # return true if ip is between (inclusive) ip_address_start and ip_address_end
  # ip_address_start and ip_address_end use CIDR notation; e.g. '192.168.1.1/24'
  # All the addresses must be passed as strings
  def host_in_range?(ip, ip_address_start, ip_address_end)
    ip >= ip_address_start && ip <= ip_address_end
  end

  def host?(hostname, ip_address)
    records = @resolver.getresources(hostname, Resolv::DNS::Resource::IN::A)
    records.each do |record|
      return true if record.address.to_s == ip_address
    end
    false
  end

  def host_ipv6?(hostname, ip_address)
    records = @resolver.getresources(hostname, Resolv::DNS::Resource::IN::AAAA)
    records.each do |record|
      return true if record.address.to_s.casecmp(ip_address)
    end
    false
  end

  def pointer?(hostname, ip_address)
    # Using dig due to:
    # https://stackoverflow.com/questions/27060993/resolvdns-not-resolving-ptr-for-192-168-x
    resolved_hostname = `/usr/bin/dig @#{@ip_address} -x #{ip_address} +short`
    return true if resolved_hostname.delete("\n") == "#{hostname}."
    false
  end

  def mail_server?(domain_name, hostname, preference)
    mx_records = @resolver.getresources(domain_name,
                                        Resolv::DNS::Resource::IN::MX)
    mx_records.each do |record|
      return true if record.exchange.to_s == hostname &&
                     record.preference == preference
    end
    false
  end

  def nameserver?(domain_name, hostname)
    ns_records = @resolver.getresources(domain_name,
                                        Resolv::DNS::Resource::IN::NS)
    ns_records.each do |record|
      STDOUT.write 'NS: ' + record.name.to_s + '==' + hostname
      return true if record.name.to_s == hostname
    end
    false
  end

  def alias?(hostname, host_alias)
    records = @resolver.getresources(hostname, Resolv::DNS::Resource::IN::CNAME)
    records.each do |record|
      return true if record.name.to_s == host_alias
    end
    false
  end
end


spec/spec_helper.rb
require 'puppetlabs_spec_helper/module_spec_helper'
require 'rspec-puppet-facts'
include RspecPuppetFacts

if Dir.exist?(File.expand_path('../../lib', __FILE__))
  require 'coveralls'
  require 'simplecov'
  require 'simplecov-console'
  SimpleCov.formatters = [
    SimpleCov::Formatter::HTMLFormatter,
    SimpleCov::Formatter::Console,
    Coveralls::SimpleCov::Formatter
  ]
  SimpleCov.start do
    track_files 'lib/**/*.rb'
    add_filter '/spec'
    add_filter '/vendor'
    add_filter '/.vendor'
  end
end

RSpec.configure do |c|
  default_facts = {
    puppetversion: Puppet.version,
    facterversion: Facter.version
  }
  default_facts.merge!(YAML.load(File.read(File.expand_path('../default_facts.yml', __FILE__)))) if File.exist?(File.expand_path('../default_facts.yml', __FILE__))
  default_facts.merge!(YAML.load(File.read(File.expand_path('../default_module_facts.yml', __FILE__)))) if File.exist?(File.expand_path('../default_module_facts.yml', __FILE__))
  c.default_facts = default_facts
end
# vim: syntax=ruby


lib/spec/nsupdate.rb
require 'resolv'

class Nsupdate
  def initialize(server, dns_key)
    @dns_key = dns_key
    @server = server
  end

  # create({ :fqdn => "node01.lab", :value => "192.168.100.2"}
  # create({ :fqdn => "node01.lab", :value => "3.100.168.192.in-addr.arpa",
  #          :type => "PTR"}
  def create(fqdn, value, type, ttl = 600, priority = 10)
    nsupdate 'connect'
    # forward_domain = URI.parse(fqdn).host

    @resolver = Resolv::DNS.new(nameserver: @server)
    case type
    when 'A'
      # nsupdate "zone #{forward_domain}"
      nsupdate "update add #{fqdn}. #{ttl} #{type} #{value}"
    when 'AAAA'
      # nsupdate "zone #{forward_domain}"
      nsupdate "update add #{fqdn}. #{ttl} #{type} #{value}"
    when 'CNAME'
      # nsupdate "zone #{forward_domain}"
      nsupdate "update add #{fqdn}. #{ttl} #{type} #{value}."
    when 'PTR'
      reverse_ip_array = fqdn.split('.').reverse
      ptr_record = reverse_ip_array.join('.') + '.IN-ADDR.ARPA'
      reverse_ip_array.shift # Remove first octet
      cnet_domain = reverse_ip_array.join('.') + '.IN-ADDR.ARPA'
      nsupdate "zone #{cnet_domain}"
      nsupdate "update add #{ptr_record}. #{ttl} #{type} #{value}."
    when 'NS'
      nsupdate "zone #{fqdn}"
      nsupdate "update add #{fqdn} 0 IN #{type} #{value}"
    when 'MX'
      # nsupdate "zone #{forward_domain}"
      nsupdate "update add #{fqdn} #{ttl} #{type} #{priority} #{value}."
    end
    nsupdate 'disconnect'
  ensure
    @om.close unless @om.nil? || @om.closed?
  end

  # remove({ :fqdn => "node01.lab", :value => "192.168.100.2"}
  def remove
    nsupdate 'connect'
    case @type
    when 'A'
      nsupdate "update delete #{@fqdn} #{@type}"
    when 'PTR'
      nsupdate "update delete #{@value} #{@type}"
    end
    nsupdate 'disconnect'
  end

  protected

  def nsupdate_args
    args = ''
    args = "-k #{@dns_key} " if @dns_key
    args
  end

  def nsupdate(cmd)
    status = nil
    if cmd == 'connect'
      find_nsupdate if @nsupdate.nil?
      nsupdate_cmd = "#{@nsupdate} #{nsupdate_args}"
      #      logger.debug "running #{nsupdate_cmd}"
      STDOUT.write "running #{nsupdate_cmd}\n"
      @om = IO.popen(nsupdate_cmd, 'r+')
      #      logger.debug "nsupdate: executed - server #{@server}"
      STDOUT.write "nsupdate: executed - server #{@server}\n"
      @om.puts "server #{@server}"
    elsif cmd == 'disconnect'
      @om.puts 'send'
      @om.puts 'answer'
      @om.close_write
      status = @om.readlines
      @om.close
      @om = nil # we cannot serialize an IO object, even if closed.
      # TODO Parse output for errors!
      if !status.empty? && status[1] !~ %r{status: NOERROR}
        #        logger.debug "nsupdate: errors\n" + status.join("\n")
        #        raise Proxy::Dns::Error.new("Update errors: #{status.join("\n")}")
        raise "Update errors: #{status.join("\n")}"
      end
    else
      STDOUT.write "nsupdate: executed - #{cmd}\n"
      #      logger.debug "nsupdate: executed - #{cmd}"
      @om.puts cmd
    end
  end

  private

  def find_nsupdate
    @nsupdate = which('nsupdate')
    raise 'unable to find nsupdate binary' unless File.exist?(@nsupdate.to_s)
  end

  def dns_find(key)
    match = key.match(%r{(\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3})})
    if match
      resolver.getname(match[1..4].reverse.join('.')).to_s
    else
      resolver.getaddress(key).to_s
    end
  rescue Resolv::ResolvError
    false
  end

  def which(cmd)
    exts = ENV['PATHEXT'] ? ENV['PATHEXT'].split(';') : ['']
    ENV['PATH'].split(File::PATH_SEPARATOR).each do |path|
      exts.each do |ext|
        exe = File.join(path, "#{cmd}#{ext}")
        return exe if File.executable?(exe) && !File.directory?(exe)
      end
    end
    nil
  end
end


lib/dns.rb
require 'rubygems'
require 'resolv'
require 'ipaddress'

class DNS

  # ip_address = dotted notation as a string, e.g. '192.168.1.1'
  # role = 'Master' or 'Slave'
  # domain_name = the domain name as a string, e.g. 'sharknet.us'
  def initialize(ip_address, role, domain_name)
    @role = role
    @ip_address = ip_address
    @resolver = Resolv::DNS.new(
      :nameserver => [ip_address],
      :search => [domain_name],
      :ndots => 1)
    # @resolver.timeouts = 2 # Enable for Ruby 2.1 or newer. Fail if lookup takes longer than 2 seconds
  end

  # returns the ip address string
  def name
    @ip_address
  end

  # returns the role string
  def role
    @role
  end

  # returns the DNS server as a string in the form '192.168.1.1 [Master]'
  def to_s
    "#{name} [#{role}]"
  end

  # returns the ip address as a string or 'Hostname not found' if there was any error
  def address( hostname )
   begin
    @resolver.getaddress( hostname ).to_s
  rescue
   'Hostname not found'
 end
end

# returns the hostname as a string or 'IP address not found' if there was any error
def hostname( ip_address )
  begin
   @resolver.getname( ip_address ).to_s
 rescue
   'IP address not found'
 end
end

  # return true if ip is between (inclusive) ip_address_start and ip_address_end
  # ip_address_start and ip_address_end use CIDR notation; e.g., '192.168.1.1/24'
  # All the addresses must be passed as strings
  def host_in_range?( ip, ip_address_start, ip_address_end )
    return ip >= ip_address_start && ip <= ip_address_end
  end

end
